### PR TITLE
AddAlgorithmTesting

### PR DESCRIPTION
**Pull request by geoffjohn11**

### CODE REVIEW

**Comment by geoffjohn11** [comment]
@BalmungSan Not entirely certain I'm constructing the docker image with the volume the best way.

File: core/src/test/scala/neotypes/algorithms/CentralitySpec.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      |MERGE (paper6)-[:CITES]->(paper4);
+    """.stripMargin
+
+  it should "execute the article rank algorithm" in execute { s =>
+    """CALL algo.articleRank('Paper', 'CITES',
+      {iterations:20, dampingFactor:0.85, write: true,writeProperty:"pagerank"})
+      YIELD nodes, iterations, dampingFactor, write, writeProperty""".query[ArticleRanking].single(s)
+      .map(_ shouldBe ArticleRanking(7, 20, 0.85, true, "pagerank"))
+  }
+}
```
Comment (BalmungSan) [inline core/src/test/scala/neotypes/algorithms/CentralitySpec.scala]: Please make sure all files end with an `EOL` before the `EOF`

File: core/src/test/scala/neotypes/algorithms/CommunityDetectionSpec.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      |MERGE (chris)-[:KNOWS]->(karin);
+    """.stripMargin
+
+  it should "execute the triangle count algorithm" in execute { s =>
+    """CALL algo.triangleCount('Person', 'KNOWS',
+      {concurrency:4, write:true, writeProperty:'triangles',clusteringCoefficientProperty:'coefficient'})
+      YIELD loadMillis, computeMillis, writeMillis, nodeCount, triangleCount, averageClusteringCoefficient;""".query[TriangleCount].single(s)
+      .map(_ shouldBe TriangleCount(6, 3, 0.6055555555555555))
+  }
+}
```
Comment (BalmungSan) [inline core/src/test/scala/neotypes/algorithms/CommunityDetectionSpec.scala]: Please make sure all files end with an `EOL` before the `EOF`.

File: core/src/test/scala/neotypes/algorithms/LinkPredictionSpec.scala [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      |MERGE (michael)-[:WORKS_WITH]-(karin)
+      |MERGE (arya)-[:FRIENDS]-(karin)""".stripMargin
+
+  it should "execute the Adamic Adar algorithm" in execute { s =>
+    """MATCH (p1:Person {name: 'Michael'})
+      MATCH (p2:Person {name: 'Karin'})
+      RETURN algo.linkprediction.adamicAdar(p1, p2) AS score""".query[AdamicLink].single(s)
+      .map( _ shouldBe AdamicLink(0.9102392266268373))
+  }
+}
```
Comment (BalmungSan) [inline core/src/test/scala/neotypes/algorithms/LinkPredictionSpec.scala]: Please make sure all files end with an `EOL` before the `EOF`.

File: core/src/test/scala/neotypes/algorithms/ShortestPathSpec.scala [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    """.stripMargin
+
+  it should "execute the shortest path algorithm" in execute { s =>
+    """MATCH (start:Loc{name:'A'}), (end:Loc{name:'F'})
+                CALL algo.shortestPath(start, end, 'cost',{write:true,writeProperty:'sssp'})
+                YIELD writeMillis,loadMillis,nodeCount, totalCost
+                RETURN writeMillis,loadMillis,nodeCount,totalCost""".query[ShortestPath].single(s)
+      .map(_ shouldBe ShortestPath(5, 160.0))
+  }
+}
```
Comment (BalmungSan) [inline core/src/test/scala/neotypes/algorithms/ShortestPathSpec.scala]: Please make sure all files end with an `EOL` before the `EOF`.

File: core/src/test/scala/neotypes/algorithms/SimilaritySpec.scala [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    """MATCH (p1:Person {name: 'Karin'})-[:LIKES]->(cuisine1)
+      WITH p1, collect(id(cuisine1)) AS p1Cuisine
+      MATCH (p2:Person {name: "Arya"})-[:LIKES]->(cuisine2)
+      WITH p1, p1Cuisine, p2, collect(id(cuisine2)) AS p2Cuisine
+      RETURN p1.name AS from,
+             p2.name AS to,
+             algo.similarity.jaccard(p1Cuisine, p2Cuisine) AS similarity""".query[JaccardSimilarity].single(s)
+      .map(_ shouldBe JaccardSimilarity("Karin", "Arya", 0.6666666666666666))
+  }
+}
```
Comment (BalmungSan) [inline core/src/test/scala/neotypes/algorithms/SimilaritySpec.scala]: Please make sure all files end with an `EOL` before the `EOF`.

File: core/src/test/scala/neotypes/BaseIntegrationSpec.scala [hunk 6]
```diff
@@ -40,6 +41,14 @@ abstract class BaseIntegrationSpec[F[_]] extends AsyncFlatSpec with ForAllTestCo
   }
 }
 
+abstract class BaseAlgorithmIntegrationSpec extends BaseIntegrationSpec[scala.concurrent.Future]{
+  override val container = GenericContainer("neo4j:3.5.3",
+    exposedPorts = Seq(7687),
+    waitStrategy = new HostPortWaitStrategy().withStartupTimeout(Duration.ofSeconds(30))
+  ).configure(_.withClasspathResourceMapping("neo4j.conf", "/var/lib/neo4j/conf/", BindMode.READ_ONLY))
+    .configure(_.withClasspathResourceMapping("graph-algorithms-algo-3.5.4.0.jar", "/var/lib/neo4j/plugins/", BindMode.READ_ONLY))
+}
```
Comment (BalmungSan) [inline core/src/test/scala/neotypes/BaseIntegrationSpec.scala]: All this should be on its own file.

**Pull request merged**